### PR TITLE
Scraper

### PR DESCRIPTION
**Pull request by ingridel**
Implemented a scraper for Google Play Store. For now, it can only be used for downloading app icons.

### CODE REVIEW

File: play_scraper.py
Comment (s-bayer) [inline play_scraper.py]: You don't need types in your docstring. Use typehints instead. They are actually enforceable on the call-site.

I would also normally not document private properties in a docstring. If you want to document them I would do it where they are passed (in the constructor in this case).
As to all rules there are of course exception for this, for instance if they are used to uphold complex invariants, but in this case it's pretty clear what they do from the name.

File: play_scraper.py
Comment (s-bayer) [inline play_scraper.py]: In Python 3 I would use `pathlib` instead of `os`. It's more succinct, type-safe and makes the intention clearer.
I would also consider requiring a `Path` as parameter to your constructor.
```suggestion
        self._storage_dir = pathlib.Path(storage_dir)
        self._storage_dir.mkdir(parents=True, exist_ok=True)
```

File: play_scraper.py
Comment (s-bayer) [inline play_scraper.py]: Typehints :)
Conveniently calling the `Path` constructor on a `Path` (instead of a `str`) is a no-op, so this just works with the changes suggested below.

```suggestion
    def __init__(self, base_url: str, storage_dir: Union[str, pathlib.Path]):
```

File: play_scraper.py
Comment (s-bayer) [inline play_scraper.py]: For private functions I would not bother with docstrings.
Everything you say in the docstring is already clear from the function name, parameter names and typehints (once you add them) and all of those are more resistant to code rot.

Pylint is [extremely configurable](http://pylint.pycqa.org/en/latest/technical_reference/features.html#basic-checker-options), so it might be possible to disable missing docstring warnings for private functions. It's definitely possible to [disable the warnings for short functions](https://stackoverflow.com/a/26889689) which might be a good alternative. For instance adding a docstring to a `get_item(self): return self.item` method makes your code harder to read instead of easier.

File: play_scraper.py
Comment (s-bayer) [inline play_scraper.py]: One underscore is almost always enough and using two [can be harmful](https://stackoverflow.com/a/1301369). It  also does not really make things more private and just mangles the name. This means someone who really wants to access it still easily can do so, it just makes the code less readable. As such it is also [discouraged in the style guide](https://google.github.io/styleguide/pyguide.html#3162-naming-conventions).

For these reasons when people say "make something private" in a Python-context they basically always refer to just one underscore.

```suggestion
    def _get_html(self, app_id, language=None):
```

File: play_scraper.py
Comment (s-bayer) [inline play_scraper.py]: Using `+` to append a subdirectory to a path is brittle and will silently break if someone does not add a trailing underscore to the parent directory. It is also not platform-independent.
For these reasons I would always use the helpers in `pathlib` or `os.path.join`.
In the suggestion I assume that we changed `storage_dir` to be a `Path` instance.
Note that `Path('a/b/c') / "" == Path('a/b/c') == Path('a/b') / 'c'`, meaning you don't need to handle the case where no `directory` was passed separately.

```suggestion
        location = self._storage_dir / directory
        location.mkdir(exist_ok=True)
        urlretrieve(src, location / f"icon_{app_id}")
```

File: test_data/icon_com.example
Comment (s-bayer) [inline test_data/icon_com.example]: Since you regenerate this file with every test run it should be in the .gitignore.

File: play_scraper_test.py
Comment (s-bayer) [inline play_scraper_test.py]: To me it looks like it could be easier (and cleaner) to use [tmpdir](https://docs.pytest.org/en/stable/tmpdir.html) to create both the html and the png dynamically in the test case.
In addition to removing the parameters you introduced in your conftest.py, this would also save you the problems with absolute vs. relative paths and would also make it clear right in the testcase what exactly is tested.

File: play_scraper.py
Comment (s-bayer) [inline play_scraper.py]: This function mixed low-level logic (how the url looks like) with high-level logic (build a url, then open the page with BeautifulSoup).
I would move the url-building to a helper function to only have a single level of abstraction in this function.

The principle behind this  is called SLA (single level of abstraction) and is described [here](http://www.principles-wiki.net/principles:single_level_of_abstraction) or in the Clean Code book. [This blog post](https://codurance.com/2015/01/27/balanced-abstraction-principle/#:~:text=The%20Balanced%20Abstraction%20Principle%20defines,the%20same%20level%20of%20abstraction) has a more succinct and probably easier to follow description.

**Review by s-bayer — CHANGES_REQUESTED** [review]
Congrats on your first pull request 🎉

File: test/play_scraper_test.py
Comment (miivanov) [inline test/play_scraper_test.py]: please add tests where the html attributes we are looking for are not present to make sure your code gracefully handles that

File: src/play_scraper.py
Comment (miivanov) [inline src/play_scraper.py]: Can we make the code resilient to cases where this attribute is missing (in case the html changes for example) it looks like your code simply breaks

File: src/play_scraper.py
Comment (miivanov) [inline src/play_scraper.py]: Same thing here -> what happens if the category cannot be found? let's make that not fail with an error

File: src/play_scraper.py
Comment (miivanov) [inline src/play_scraper.py]: Please put that into a method called build_app_details_page_url (or similar) and use the same method below

File: src/play_scraper.py
Comment (miivanov) [inline src/play_scraper.py]:
```suggestion
        """Scrapes the app icon URL from the app's play store details page,
        downloads the corresponding app icon, and saves it to the specified subdirectory.

        :param app_id: the id of the app.
        :param directory: icon storage subdirectory.
```

Comment (miivanov) [inline src/play_scraper.py]: Clarify that this subdirectory is appended to the base storage directory provided.

File: src/play_scraper.py
Comment (miivanov) [inline src/play_scraper.py]: Similarly clarify that the app's category is scraped from the app's play store details page.

File: src/play_scraper.py
Comment (miivanov) [inline src/play_scraper.py]: this method is quite involved, let's split it into:
-> get_app_details_page(app_id) (returns the app details page html contents)
-> scrape_app_icon_url(html) (scrapes the app icon url given a page)
-> download_app_icon(icon_url, directory) (creates the directory and downloads&saves the image there)
-> download_app_icon(app_id, directory) (same as your original method)

By splitting the methods you'd be able to easily test individual logic (and corner cases like failed requests), improve readability, and more clearly handle failures due to different methods returning unexpected results.

**Review by miivanov — COMMENTED** [review]
Looking good, needs a bit more work though, thanks!

The summary of my comments is:
-> splitting the current API into smaller logical chunks
-> handling error cases (html page is empty, downloading failed, the html attribute is missing)
-> increasing test coverage to address failure cases

File: src/play_scraper.py
Comment (ingridel) [reply src/play_scraper.py]: I see that the last two suggested functions have the same name. Can you please elaborate on it? Do you want to overload this method or is it just a mistake?

Comment (miivanov) [reply src/play_scraper.py]: Whoops, sorry about the confusion, as per our discussion I meant the download_app_icon(icon_url, directory) to be download_img(img_url, directory)

File: src/play_scraper.py
Comment (miivanov) [inline src/play_scraper.py]: Add an explicit comment that the app icon is stored at storage_dir/sub_dir/icon_{app_id}

File: src/play_scraper.py
Comment (miivanov) [inline src/play_scraper.py]: nit: rename directory to sub_dir to explicitly hint at the purpose and perhaps give it a default value: "icons" or whatever you've chosen

File: test/play_scraper_test.py
Comment (miivanov) [inline test/play_scraper_test.py]: do you need rand here? :) Just manually hardcode something

File: test/play_scraper_test.py
Comment (miivanov) [inline test/play_scraper_test.py]: again -> do we really need rand here -> let's just have a simple hardcoded array to improve visibility and drop dependency on np for the test.

**Review by miivanov — CHANGES_REQUESTED** [review]
LGTM from my side - just resolve the few small nits I've left and wait for Seb's approval please :)

File: test/play_scraper_test.py
Comment (s-bayer) [reply test/play_scraper_test.py]: Also 10000 is too small and could still cause a collision. I would use something like `betel-test-invalid-base-url` to make it specific to your codebase which makes collisions much less likely. If you want to be extra sure `betel-test-invalid-base-url-938TD`

File: src/play_scraper.py
Comment (s-bayer) [inline src/play_scraper.py]: I would also mention what you scrape (either everything or examples).

File: src/play_scraper.py
Comment (s-bayer) [inline src/play_scraper.py]: Currently you first define all helpers and then use them in your methods (private and public ones). I would exactly reverse this order to let people quickly see the public functions in you class and to make it easier to scan code from a high level of abstraction towards a low level of abstraction (implementation details).

This also goes for your private helpers on a module level. I would move them below the class so that someone opening this file sees your public class (and important constants, if any) first.

File: src/play_scraper.py
Comment (s-bayer) [inline src/play_scraper.py]: I would make sure that all your exceptions inherit from your own base exception (e.g. BetelException) which itself of course inherits from `Exception`.

[This article](https://julien.danjou.info/python-exceptions-guide/) covers pretty well how you should deal with exceptions in Python (and in any language, really).

File: test/conftest.py
Comment (s-bayer) [inline test/conftest.py]: Your tests should be independent of order and you should not reuse the same test directory for multiple tests, so I would remove the `scope` here and in `icon_dir`

File: test/conftest.py
Comment (s-bayer) [inline test/conftest.py]: As mentioned above, I would normally use no scope here. The only exception would be if your tests would run unreasonably long, which I don't think they do.

File: test/play_scraper_test.py
Comment (s-bayer) [inline test/play_scraper_test.py]: I would move the test file creation (html and icons) into high-level helpers to make the tests easier to read.

File: test/play_scraper_test.py
Comment (s-bayer) [inline test/play_scraper_test.py]: If you don't reuse your test dir (by not setting a scope), those filenames can always be the same, making your tests slightly easier to read.

**Review by s-bayer — APPROVED** [review]
Tentatively accepted. Please resolve all comments before merging.

File: src/play_scraper.py
Comment (ingridel) [reply src/play_scraper.py]: I was thinking that the default should be downloading in the storage_dir, without any subdirectory. Is it ok?

**Review by miivanov — APPROVED** [review]

**Pull request merged**
